Extractor with structural type crashes pattern matching

## Minimized code

```Scala
class Nat(val x: Int) {
  def get: Int = x
  def isEmpty = x < 0
}

val SomeExtractorBuilder: { def unapply(x: Int): Nat } = new {
  def unapply(x: Int): Nat = new Nat(x)
}

5 match {
  case SomeExtractorBuilder(n) => println(s"$n is a natural number")
  case _      => ()
}
```

## Output (click arrow to expand)
<details>

```scala
Exception in thread "main" java.lang.AssertionError: assertion failed
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:16)
	at dotty.tools.dotc.typer.Applications$.unapplyArgs(Applications.scala:173)
	at dotty.tools.dotc.typer.Applications.typedUnApply(Applications.scala:1177)
	at dotty.tools.dotc.typer.Typer.typedUnApply(Typer.scala:83)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2119)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2177)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2213)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2225)
	at dotty.tools.dotc.typer.Typer.typedPattern(Typer.scala:2337)
	at dotty.tools.dotc.typer.Typer.typedCase(Typer.scala:1217)
	at dotty.tools.dotc.typer.Typer.typedCases$$anonfun$1(Typer.scala:1174)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:63)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:79)
	at dotty.tools.dotc.typer.Typer.typedCases(Typer.scala:1176)
	at dotty.tools.dotc.typer.Typer.$anonfun$18(Typer.scala:1166)
	at dotty.tools.dotc.typer.Applications.harmonic(Applications.scala:1915)
	at dotty.tools.dotc.typer.Typer.harmonic(Typer.scala:83)
	at dotty.tools.dotc.typer.Typer.typedMatchFinish(Typer.scala:1166)
	at dotty.tools.dotc.typer.Typer.typedMatch(Typer.scala:1148)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2132)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2177)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2213)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2225)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1263)
	at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1253)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1263)
	at dotty.tools.dotc.typer.Namer.rhsType$2(Namer.scala:1397)
	at dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1408)
	at dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1409)
	at dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1420)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1428)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:811)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:935)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:843)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:258)
	at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:185)
	at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:187)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:398)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2071)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2096)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2176)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2213)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2225)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2247)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2291)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1784)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2109)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2176)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2213)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2225)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2247)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2291)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1910)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2150)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2177)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2213)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2225)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2333)
	at dotty.tools.dotc.typer.FrontEnd.liftedTree1$2(FrontEnd.scala:78)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:83)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:42)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:84)
	at dotty.tools.repl.REPLFrontEnd.runOn(ReplFrontEnd.scala:23)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:161)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:171)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:179)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:65)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:186)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:128)
	at dotty.tools.repl.ReplCompiler.runCompilationUnit(ReplCompiler.scala:156)
	at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:166)
	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:229)
	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:193)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:127)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:130)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:148)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:130)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/reporting/messages.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/messages.scala
@@ -1806,7 +1806,7 @@ object messages {
         |""".stripMargin
   }
 
-  class UnapplyInvalidReturnType(unapplyResult: Type, unapplyName: Symbol#ThisName)(implicit ctx: Context)
+  class UnapplyInvalidReturnType(unapplyResult: Type, unapplyName: Name)(implicit ctx: Context)
     extends DeclarationMsg(UnapplyInvalidReturnTypeID) {
     def msg =
       val addendum =

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -150,7 +150,10 @@ object Applications {
 
   def unapplyArgs(unapplyResult: Type, unapplyFn: Tree, args: List[untpd.Tree], pos: SourcePosition)(using Context): List[Type] = {
 
-    val unapplyName = unapplyFn.symbol.name
+    val unapplyName = unapplyFn match            // tolerate structural `unapply`, which does not have a symbol
+      case TypeApply(fn: RefTree, _) => fn.name
+      case fn: RefTree => fn.name
+
     def getTp = extractorMemberType(unapplyResult, nme.get, pos)
 
     def fail = {
@@ -1227,6 +1230,9 @@ trait Applications extends Compatibility {
           case Apply(Apply(unapply, `dummyArg` :: Nil), args2) => assert(args2.nonEmpty); args2
           case Apply(unapply, `dummyArg` :: Nil) => Nil
           case Inlined(u, _, _) => unapplyImplicits(u)
+          case DynamicUnapply(_) =>
+            ctx.error("Structural unapply is not supported", unapplyFn.sourcePos)
+            Nil
           case _ => Nil.assertingErrorsReported
         }
 

--- a/compiler/src/dotty/tools/dotc/typer/Dynamic.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Dynamic.scala
@@ -22,6 +22,17 @@ object Dynamic {
     name == nme.applyDynamic || name == nme.selectDynamic || name == nme.updateDynamic || name == nme.applyDynamicNamed
 }
 
+object DynamicUnapply {
+  def unapply(tree: tpd.Tree): Option[List[tpd.Tree]] = tree match
+    case TypeApply(Select(qual, name), _) if name == nme.asInstanceOfPM =>
+      unapply(qual)
+    case Apply(Apply(Select(selectable, fname), Literal(Constant(name)) :: ctag :: Nil), _ :: implicits)
+    if fname == nme.applyDynamic && (name == "unapply" || name == "unapplySeq") =>
+      Some(selectable :: ctag :: implicits)
+    case _ =>
+      None
+}
+
 /** Handles programmable member selections of `Dynamic` instances and values
  *  with structural types. Two functionalities:
  *

--- a/tests/neg/i8730.scala
+++ b/tests/neg/i8730.scala
@@ -1,0 +1,18 @@
+import reflect.Selectable.reflectiveSelectable
+
+class Nat(val x: Int) {
+  def get: Int = x
+  def isEmpty = x < 0
+}
+
+val SomeExtractorBuilder: { def unapply(x: Int): Nat } = new {
+  def unapply(x: Int): Nat = new Nat(x)
+}
+
+
+@main
+def Test = 5 match {
+  case SomeExtractorBuilder(n) => println(s"$n is a natural number") // error
+  case _      => ()
+}
+

